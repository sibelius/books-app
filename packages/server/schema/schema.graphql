input MeEditInput {
  """User name. ex: Jean"""
  name: String

  """Language of the user. ex: pt"""
  lang: String
  clientMutationId: String
}

type MeEditPayload {
  me: User

  """Default error field resolver for mutations"""
  error: String
  clientMutationId: String
}

type Mutation {
  UserLogin(input: UserLoginInput!): UserLoginPayload
  UserRegistration(input: UserRegistrationInput!): UserRegistrationPayload
  MeEdit(input: MeEditInput!): MeEditPayload
  UserForgotPassword(input: UserForgotPasswordInput!): UserForgotPasswordPayload
}

"""An object with an ID"""
interface Node {
  """The id of the object."""
  id: ID!
}

"""The root of all... queries"""
type Query {
  """The ID of an object"""
  id: ID!

  """Fetches an object given its ID"""
  node(
    """The ID of an object"""
    id: ID
  ): Node

  """Fetches objects given their IDs"""
  nodes(
    """The IDs of objects"""
    ids: [ID!]!
  ): [Node]!

  """Me is the logged User"""
  me: User
  status: Status
}

"""Status"""
type Status {
  """Books app environment"""
  env: String

  """Server build version"""
  version: String

  """Server commit sha, GITHUB_SHA"""
  commitSha: String

  """Books app environment"""
  e: String

  """Server build version"""
  v: String

  """Server commit sha, GITHUB_SHA"""
  c: String
}

"""Represents an user"""
type User implements Node {
  """The ID of an object"""
  id: ID!

  """MongoDB _id"""
  _id: String!

  """User name resolver"""
  name: String

  """User surname resolver"""
  surname: String

  """User name resolver"""
  fullName: String

  """User email resolver"""
  email: String

  """User email resolver"""
  emailWasVerified: String
  lang: String

  """The soft delete status."""
  isActive: Boolean

  """The hard delete status."""
  removedAt: String

  """The date that the node was created."""
  createdAt: String

  """The date that the node was last updated."""
  updatedAt: String
}

input UserForgotPasswordInput {
  email: String!
  clientMutationId: String
}

type UserForgotPasswordPayload {
  """Default error field resolver for mutations"""
  error: String
  clientMutationId: String
}

input UserLoginInput {
  """User email to be used on login. ex: jean@gmail.com"""
  email: String!

  """User password."""
  password: String!
  clientMutationId: String
}

type UserLoginPayload {
  token: String

  """Default error field resolver for mutations"""
  error: String
  clientMutationId: String
}

input UserRegistrationInput {
  """User name. ex: Jean"""
  name: String!

  """User email to be used on login. ex: jean@gmail.com"""
  email: String!

  """User password."""
  password: String!
  clientMutationId: String
}

type UserRegistrationPayload {
  token: String

  """Default error field resolver for mutations"""
  error: String
  clientMutationId: String
}
