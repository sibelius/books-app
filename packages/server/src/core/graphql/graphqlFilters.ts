// inspired by: https://github.com/entria/graphql-mongo-helpers

import { ConnectionArguments } from 'graphql-relay';
import idx from 'idx';

import { JSObject } from '@booksapp/types';

import { GraphQLContext } from '../../types';
import { getObjectId } from '../../common/utils';

import { DIRECTION } from './enum/DirectionEnumType';

export const FILTER_CONDITION_TYPE = {
  MATCH_1_TO_1: 'MATCH_1_TO_1', // something that could be used on find() or $match
  CUSTOM_CONDITION: 'CUSTOM_CONDITION', // create a custom condition based on value
  AGGREGATE_PIPELINE: 'AGGREGATE_PIPELINE', // just an aggregate
};

export type BuiltConditionSet = {
  conditions: JSObject;
  pipeline: JSObject[];
};

type FilterFieldMappingMatch<ValueT> = {
  type: typeof FILTER_CONDITION_TYPE.MATCH_1_TO_1;
  key: string;
  format?: (value: ValueT, filter: JSObject, context: GraphQLContext, options?: JSObject<any>) => any;
};

type FilterFieldMappingPipeline<ValueT> = {
  type: typeof FILTER_CONDITION_TYPE.AGGREGATE_PIPELINE;
  pipeline: JSObject[] | ((value: ValueT) => Array<{}>);
};

type FilterFieldMappingCustomCondition<ValueT> = {
  type: typeof FILTER_CONDITION_TYPE.CUSTOM_CONDITION;
  format: (value: ValueT, filter: JSObject, context: GraphQLContext, options?: JSObject<any>) => any;
};

type FilterFieldMapping<ValueT> =
  | FilterFieldMappingMatch<ValueT>
  | FilterFieldMappingPipeline<ValueT>
  | FilterFieldMappingCustomCondition<ValueT>
  | boolean;

export type FilterMapping = { [key: string]: FilterFieldMapping<any> };

const validOperators = ['gt', 'gte', 'lt', 'lte', 'in', 'nin', 'ne', 'all'];

const arrayOperators = ['in', 'nin', 'all'];

const getFilterName = (filterName: string) => filterName.split('_')[0];

type Operators = 'gt' | 'gte' | 'lt' | 'lte' | 'in' | 'nin' | 'ne' | 'all' | '$and' | '$or';

const handleAndOr = (operator: Operators) => (
  context: GraphQLContext,
  condition: Array<{}>,
  mapping: { [key: string]: FilterFieldMapping<{}> },
) => {
  if (!Array.isArray(condition)) {
    throw new Error(`Invalid filter supplied to ${operator}.`);
  }

  return {
    condition: condition.map((andCondition) => buildConditionsObject(context, andCondition, mapping)),
    conditionName: operator,
  };
};

const handleAnd = handleAndOr('$and');
const handleOr = handleAndOr('$or');

const handleFieldOperator = (
  context: GraphQLContext,
  condition: JSObject,
  conditionName: string,
  fieldMapping: FilterFieldMapping<JSObject>,
  prev: JSObject,
  filters: JSObject,
  options = {},
) => {
  // { "myField_operator": "something" } becomes { "myField": { $operator: "something" } }
  // { "myField": "something" } remains the same
  const conditionNamePieces = conditionName.split('_');
  const operator = conditionNamePieces.length > 1 ? conditionNamePieces.pop() : '';
  // I don't think we support snake case for field names, should this be here?
  conditionName = conditionNamePieces.join('_');

  if (fieldMapping && fieldMapping.format && typeof fieldMapping.format === 'function') {
    condition = fieldMapping.format(condition, filters, context, options);
  }

  if (operator) {
    if (validOperators.indexOf(operator) === -1) {
      throw new Error(`"${operator}" is not a valid operator on field "${conditionName}".`);
    }

    if (arrayOperators.indexOf(operator) >= 0 && !Array.isArray(condition)) {
      throw new Error(`Field "${conditionName}" must have an array value.`);
    }

    condition = {
      [`$${operator}`]: condition,
    };
  }

  // handle $gte and $let fields merge
  if (conditionName in prev) {
    condition = {
      ...condition,
      ...prev[conditionName],
    };
  }

  return {
    condition,
    conditionName,
  };
};

function buildConditionsObject<ValueT>(
  context: GraphQLContext,
  conditions: JSObject,
  mapping: { [key: string]: FilterFieldMapping<ValueT> },
  filters?: JSObject,
  options = {},
) {
  return Object.keys(conditions).reduce((prev, currentKey) => {
    let condition = conditions[currentKey];
    let conditionName = currentKey;

    const fieldMapping = mapping[getFilterName(currentKey)];

    if (fieldMapping === false) {
      return prev;
    }

    if (
      fieldMapping &&
      fieldMapping.type !== FILTER_CONDITION_TYPE.MATCH_1_TO_1 &&
      fieldMapping.type !== FILTER_CONDITION_TYPE.CUSTOM_CONDITION
    ) {
      return prev;
    }

    if (conditionName === 'AND') {
      ({ condition, conditionName } = handleAnd(context, condition, mapping));
    } else if (conditionName === 'OR') {
      ({ condition, conditionName } = handleOr(context, condition, mapping));
    } else {
      if (fieldMapping && fieldMapping.type === FILTER_CONDITION_TYPE.CUSTOM_CONDITION) {
        if (fieldMapping.format && typeof fieldMapping.format === 'function') {
          const customCondition = fieldMapping.format(condition, filters, context, options);

          return {
            ...prev,
            ...customCondition,
          };
        }
      }
      ({ condition, conditionName } = handleFieldOperator(
        context,
        condition,
        conditionName,
        fieldMapping,
        prev,
        filters,
        options,
      ));
    }

    conditionName = fieldMapping && fieldMapping.key ? fieldMapping.key : conditionName;

    if (condition === undefined) {
      return prev;
    }

    return {
      ...prev,
      [conditionName]: condition,
    };
  }, {});
}

export function buildMongoConditionsFromFilters<ValueT>(
  context: GraphQLContext,
  filters: JSObject,
  mapping: { [key: string]: FilterFieldMapping<ValueT> } = {},
  options = {},
): BuiltConditionSet {
  if (!filters) {
    return { conditions: {}, pipeline: [] };
  }

  const keys = Object.keys(filters);

  // first check if there are any pipeline mapped fields
  // and if AND or OR are also passed, if that is the case, we must throw an error
  // because we cannot use OR/AND while also using pipeline.
  const hasPipelineFilter = keys.find(
    (item) => idx(mapping[item], (_) => _.type) === FILTER_CONDITION_TYPE.AGGREGATE_PIPELINE,
  );
  if (hasPipelineFilter && (filters.AND || filters.OR)) {
    throw new Error(
      `Wrong filter usage, because filter "${hasPipelineFilter}" is a pipeline filter, which should disable AND and OR`,
    );
  }

  // separate filters by type
  const filtersKeysGrouped = Object.keys(filters).reduce(
    (prev, key) => {
      const filterName = getFilterName(key);

      const type = (mapping && idx(mapping[filterName], (_) => _.type)) || FILTER_CONDITION_TYPE.MATCH_1_TO_1;

      return {
        ...prev,
        [type]: {
          ...prev[type],
          [key]: filters[key],
        },
      };
    },
    {
      // start with sane defaults
      [FILTER_CONDITION_TYPE.MATCH_1_TO_1]: {},
      [FILTER_CONDITION_TYPE.CUSTOM_CONDITION]: {},
      [FILTER_CONDITION_TYPE.AGGREGATE_PIPELINE]: {},
    },
  );

  // first build our conditions JSObject.
  const conditions = buildConditionsObject(
    context,
    { ...filtersKeysGrouped.MATCH_1_TO_1, ...filtersKeysGrouped.CUSTOM_CONDITION },
    mapping,
    filters,
    options,
  );

  // now build the pipeline, which is more straightforward
  const pipeline = Object.keys(filtersKeysGrouped.AGGREGATE_PIPELINE).reduce((prev, key) => {
    const mappedFilter = mapping[key];
    // should not really happen!
    if (!mappedFilter || mappedFilter.type !== FILTER_CONDITION_TYPE.AGGREGATE_PIPELINE) {
      return prev;
    }

    const fieldPipeline = Array.isArray(mappedFilter.pipeline)
      ? mappedFilter.pipeline
      : mappedFilter.pipeline(filters[key], filters, context, options);

    return [...prev, ...(Array.isArray(fieldPipeline) ? fieldPipeline : [])];
  }, []);

  return {
    conditions,
    pipeline,
  };
}

type OrderByArg<SortT> = {
  sort: SortT;
  direction: typeof DIRECTION;
};

export function buildSortFromOrderByArg<Sort>(orderByArg: Array<OrderByArg<string>>): { [key: string]: 1 | -1 } {
  return orderByArg.reduce(
    (acc, item) => ({
      ...acc,
      [item.sort]: item.direction,
    }),
    {},
  );
}

// @TODO - treat when there's more than one $match
export function orderAggregatePipeline(pipeline: JSObject[]): JSObject[] {
  if (!pipeline.length) {
    return [];
  }

  // @TODO - support $geoWithin as well
  const geoNearStage = pipeline.find((stage) => !!stage.$geoNear);

  let orderedPipeline: JSObject[] = [];
  let conditions = {};
  for (const stage of pipeline) {
    // IMPORTANT: if the stage is geoNear put it as first of the pipeline
    if (!!stage.$geoNear) {
      orderedPipeline = [stage, ...orderedPipeline];
      continue;
    }

    // store the first appearing $match on conditions
    // discard it from the ordered pipeline
    if (!!stage.$match) {
      if (geoNearStage && Object.keys(conditions).length === 0) {
        conditions = { ...conditions, ...stage.$match };
        continue;
      }
    }

    orderedPipeline = [...orderedPipeline, stage];
  }

  // add conditions on query of $geoNear
  if (geoNearStage) {
    geoNearStage.$geoNear.query = { ...(geoNearStage.$geoNear.query || {}), ...conditions };
  }

  return orderedPipeline;
}

type ArgsWithFilter = {
  filters: { [key: string]: string };
} & { [key: string]: string } & ConnectionArguments;
export const withFilter = (args: ArgsWithFilter, filters: JSObject) => {
  return {
    ...args,
    filters: {
      ...(args.filters || {}),
      ...filters,
    },
  };
};

export const getFieldsOneToOne = (fields: string[], JSObjectId = false) => {
  const formatParam = JSObjectId ? { format: (value?: string) => value && getObjectId(value) } : {};

  return fields.reduce(
    (acc, field) => ({
      ...acc,
      [field]: {
        type: FILTER_CONDITION_TYPE.MATCH_1_TO_1,
        key: field,
        ...formatParam,
      },
    }),
    {},
  );
};
