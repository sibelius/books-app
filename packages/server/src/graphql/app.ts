/* eslint-disable no-console */
import 'isomorphic-fetch';

import { koaPlayground } from 'graphql-playground-middleware';
import { print } from 'graphql/language';
import Koa from 'koa';
import bodyParser from 'koa-bodyparser';
import convert from 'koa-convert';
import cors from '@koa/cors';
import graphqlHttp, { OptionsData } from 'koa-graphql';
import koaLogger from 'koa-logger';
import multer from 'koa-multer';
import Router from '@koa/router';
import idx from 'idx';
import i18next from 'i18next';
import koaI18next from 'koa-i18next';
import NoIntrospection from 'graphql-disable-introspection';

import * as graphqlLoaders from '../loader';

import { JWT_KEY, DEBUG_GRAPHQL } from '../common/config';

import { GraphQLContext, KoaContextExt } from '../types';

import UserModel, { IUser } from '../modules/user/UserModel';

import { t } from '../locales/helpers';

import appPlatformMiddleware from '../modules/auth/appPlatformMiddleware';
import authMiddleware from '../modules/auth/authMiddleware';

import { getDataloaders, hasIntrospectionQuery } from './helper';
import { schema } from './schema';
import { i18nMiddleware } from './i18n';

const app = new Koa<any, KoaContextExt>();
// https://github.com/koajs/koa/blob/422e539e8989e65ba43ecc39ddbaa3c4f755d465/docs/api/request.md#requestip
if (process.env.NODE_ENV === 'production') {
  app.proxy = true;
}
app.keys = [JWT_KEY];

const loaders = graphqlLoaders;

const router = new Router<any, KoaContextExt>();
const storage = multer.memoryStorage();

// https://github.com/expressjs/multer#limits
const limits = {
  // Increasing max upload size to 30 mb, since busboy default is only 1 mb
  fieldSize: 30 * 1024 * 1024,
};

app.use(bodyParser());

// https://github.com/koajs/koa/wiki/Error-Handling
// When using `koa-graphql` this is not really called, but let's keep in in here
// for future reference
// see https://github.com/chentsulin/koa-graphql/issues/85
app.use(async (ctx, next) => {
  try {
    await next();
  } catch (err) {
    console.log('koa error:', err);
    ctx.status = err.status || 500;
    ctx.app.emit('error', err, ctx);
  }
});

app.on('error', (error) => {
  console.error('Error while answering request', { error });
});

if (process.env.NODE_ENV !== 'test') {
  app.use(koaLogger());
}

app.use(convert(cors({ maxAge: 86400, origin: '*' })));

router.get('/healthz', async (ctx) => {
  try {
    await UserModel.find().lean<IUser>();
    ctx.body = 'Database working';
    ctx.status = 200;
  } catch (err) {
    ctx.body = err.toString();
    ctx.status = 500;
  }
});

router.all('/graphql', multer({ storage, limits }).any());

if (process.env.NODE_ENV !== 'production') {
  router.all(
    '/playground',
    koaPlayground({
      endpoint: '/graphql',
    }),
  );
}

// Middleware to get dataloaders
app.use(async (ctx, next) => {
  ctx.dataloaders = getDataloaders(loaders);
  await next();
});

app.use(
  koaI18next(i18next, {
    lookupCookie: 'lang', // detecting language in cookie
    order: ['cookie', 'header'],
    next: true,
  }),
);

app.use(i18nMiddleware);

// avoid requests from unknown clients
if (process.env.NODE_ENV !== 'development') {
  app.use(appPlatformMiddleware);
}

// Middleware to get result from authorization token
app.use(authMiddleware);

router.all(
  '/graphql',
  convert(
    graphqlHttp(
      async (request, ctx, koaContext): Promise<OptionsData> => {
        const { dataloaders, user } = koaContext;
        const { authorization, appversion, appbuild, appplatform, timezone = 'America/Sao_Paulo' } = request.header;

        if (process.env.NODE_ENV !== 'test') {
          console.info('Handling request', {
            appversion,
            appbuild,
            appplatform,
            timezone,
            unauthorized: !user,
            authorization,
            userId: idx(user, (_) => _._id),
          });
        }

        return {
          graphiql: process.env.NODE_ENV === 'development',
          schema,
          rootValue: { request: ctx.req },
          validationRules: process.env.NODE_ENV === 'production' ? [NoIntrospection] : [],
          context: {
            user,
            dataloaders,
            appplatform,
            koaContext,
            timezone,
            t,
          } as GraphQLContext,
          extensions: ({ document, variables, result }) => {
            if (process.env.NODE_ENV === 'development' && DEBUG_GRAPHQL) {
              if (document && !hasIntrospectionQuery(document)) {
                console.log(print(document));

                console.log(variables);

                console.log(JSON.stringify(result, null, 2));
              }
            }
            // the type of extensions does not accept null, but the code does
            return null as any;
          },
          formatError: (error: any) => {
            if (error.name && error.name === 'BadRequestError') {
              ctx.status = 400;
              ctx.body = 'Bad Request';
              return {
                message: 'Bad Request',
              };
            }

            if (error.path || error.name !== 'GraphQLError') {
              console.error(error);
            } else {
              console.log(`GraphQLWrongQuery: ${error.message}`);
            }

            console.error('GraphQL Error', { error });

            if (process.env.NODE_ENV !== 'production') {
              return {
                message: error.message,
                locations: error.locations,
                stack: error.stack,
              };
            } else {
              ctx.status = 400;
              ctx.body = 'Bad Request';
              return {
                message: 'Bad Request',
              };
            }
          },
        };
      },
    ),
  ),
);

app.use(router.routes()).use(router.allowedMethods());

export default app;
