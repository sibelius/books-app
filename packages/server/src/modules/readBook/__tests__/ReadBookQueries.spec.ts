import MockDate from 'mockdate';
import { graphql } from 'graphql';
import { toGlobalId } from 'graphql-relay';

import { sanitizeTestObject } from '@booksapp/test-utils';

import { schema } from '../../../graphql/schema';

import {
  getContext,
  clearDbAndRestartCounters,
  connectMongoose,
  disconnectMongoose,
  createUser,
  createReadBook,
  resetRunningDate,
  createBook,
  gql,
} from '../../../../test/helpers';

beforeAll(connectMongoose);

beforeEach(async () => {
  await clearDbAndRestartCounters();
  MockDate.reset();
  resetRunningDate();
  jest.clearAllMocks();
});

afterAll(disconnectMongoose);

describe('ReadBook queries', () => {
  it('should get read book from node interface', async () => {
    const user = await createUser();
    const readBook = await createReadBook();

    const query = gql`
      query Q($id: ID!) {
        readBook: node(id: $id) {
          ... on ReadBook {
            id
            readPages
            book {
              name
            }
          }
        }
      }
    `;

    const rootValue = {};
    const context = await getContext({ user });
    const variables = {
      id: toGlobalId('ReadBook', readBook._id),
    };

    const result = await graphql(schema, query, rootValue, context, variables);

    expect(result.errors).toBeUndefined();
    expect(result.data?.readBook).not.toBe(null);
    expect(sanitizeTestObject(result.data)).toMatchSnapshot();
  });

  it('should not get read book without user', async () => {
    const readBook = await createReadBook();

    const query = gql`
      query Q($id: ID!) {
        readBook: node(id: $id) {
          ... on ReadBook {
            readPages
            book {
              name
            }
          }
        }
      }
    `;

    const rootValue = {};
    const context = await getContext();
    const variables = {
      id: toGlobalId('ReadBook', readBook._id),
    };

    const result = await graphql(schema, query, rootValue, context, variables);

    expect(result.errors).toBeUndefined();
    expect(result.data?.readBook).toBe(null);
  });

  it('should get null read book if isActive is false', async () => {
    const user = await createUser();
    const event = await createReadBook({ isActive: false });

    const query = gql`
      query Q($id: ID!) {
        readBook: node(id: $id) {
          ... on ReadBook {
            readPages
            book {
              name
            }
          }
        }
      }
    `;

    const rootValue = {};
    const context = await getContext({ user });
    const variables = {
      id: toGlobalId('ReadBook', event._id),
    };

    const result = await graphql(schema, query, rootValue, context, variables);

    expect(result.errors).toBeUndefined();
    expect(result.data?.readBook).toBe(null);
  });

  it('should get null read book if removedAt exists', async () => {
    const user = await createUser();
    const event = await createReadBook({ removedAt: new Date() });

    const query = gql`
      query Q($id: ID!) {
        readBook: node(id: $id) {
          ... on ReadBook {
            readPages
            book {
              name
            }
          }
        }
      }
    `;

    const rootValue = {};
    const context = await getContext({ user });
    const variables = {
      id: toGlobalId('ReadBook', event._id),
    };

    const result = await graphql(schema, query, rootValue, context, variables);
    expect(result.errors).toBeUndefined();
    expect(result.data?.readBook).toBe(null);
  });

  it('should query a connection of read books', async () => {
    const user = await createUser();

    for (let i = 0; i < 6; i++) {
      await createReadBook();
    }

    const query = gql`
      query Q {
        readBooks(first: 5) {
          edges {
            node {
              id
              readPages
              book {
                name
              }
            }
          }
        }
      }
    `;

    const rootValue = {};
    const context = await getContext({ user });
    const variables = {};

    const result = await graphql(schema, query, rootValue, context, variables);

    expect(result.errors).toBeUndefined();
    expect(result.data?.readBooks).not.toBe(null);
    expect(result.data?.readBooks.edges.length).toBe(5);
    expect(sanitizeTestObject(result.data)).toMatchSnapshot();
  });

  it('should query a null connection of read books with there is no user on ctx', async () => {
    for (let i = 0; i < 6; i++) {
      await createReadBook();
    }

    const query = gql`
      query Q {
        readBooks(first: 5) {
          edges {
            node {
              id
              readPages
              book {
                name
              }
            }
          }
        }
      }
    `;

    const rootValue = {};
    const context = await getContext();
    const variables = {};

    const result = await graphql(schema, query, rootValue, context, variables);

    expect(result.errors).toBeUndefined();
    expect(result.data?.readBooks).not.toBe(null);
    expect(result.data?.readBooks.edges.length).toBe(0);
  });

  it('should query a connection of read books with orderBy filter', async () => {
    const user = await createUser();

    const readBook1 = await createReadBook({ readPages: 5 });
    const readBook2 = await createReadBook({ readPages: 3 });
    const readBook3 = await createReadBook({ readPages: 200 });

    const query = gql`
      query Q($filters: ReadBookFilters) {
        readBooks(filters: $filters) {
          edges {
            node {
              id
              readPages
              book {
                name
              }
            }
          }
        }
      }
    `;

    const rootValue = {};
    const context = await getContext({ user });
    const variables = {
      filters: {
        orderBy: { sort: 'CREATED_AT', direction: 'ASC' },
      },
    };

    const result = await graphql(schema, query, rootValue, context, variables);

    expect(result.errors).toBeUndefined();
    expect(result.data?.readBooks).not.toBe(null);
    expect(result.data?.readBooks.edges.length).toBe(3);
    expect(result.data?.readBooks.edges[0].node.readPages).toBe(readBook1.readPages);
    expect(result.data?.readBooks.edges[1].node.readPages).toBe(readBook2.readPages);
    expect(result.data?.readBooks.edges[2].node.readPages).toBe(readBook3.readPages);
    expect(sanitizeTestObject(result.data)).toMatchSnapshot();
  });

  it('should query a connection of finished read books', async () => {
    const user = await createUser();

    for (let i = 1; i < 4; i++) {
      const book = await createBook({ pages: i * 2 });
      await createReadBook({ bookId: book._id, readPages: i });
    }

    for (let i = 1; i < 11; i++) {
      const book = await createBook({ pages: i });
      await createReadBook({ bookId: book._id, readPages: i });
    }

    const query = gql`
      query Q($filters: ReadBookFilters) {
        readBooks(filters: $filters) {
          edges {
            node {
              id
              readPages
              book {
                name
              }
            }
          }
        }
      }
    `;

    const rootValue = {};
    const context = await getContext({ user });
    const variables = {
      filters: {
        finished: true,
      },
    };

    const result = await graphql(schema, query, rootValue, context, variables);

    expect(result.errors).toBeUndefined();
    expect(result.data?.readBooks).not.toBe(null);
    expect(result.data?.readBooks.edges.length).toBe(10);
    expect(sanitizeTestObject(result.data)).toMatchSnapshot();
  });

  it('should query a connection of unfinished read books', async () => {
    const user = await createUser();

    for (let i = 1; i < 7; i++) {
      const book = await createBook({ pages: i * 2 });
      await createReadBook({ bookId: book._id, readPages: i });
    }

    for (let i = 1; i < 3; i++) {
      const book = await createBook({ pages: i });
      await createReadBook({ bookId: book._id, readPages: i });
    }

    const query = gql`
      query Q($filters: ReadBookFilters) {
        readBooks(filters: $filters) {
          edges {
            node {
              id
              readPages
              book {
                name
              }
            }
          }
        }
      }
    `;

    const rootValue = {};
    const context = await getContext({ user });
    const variables = {
      filters: {
        finished: false,
      },
    };

    const result = await graphql(schema, query, rootValue, context, variables);

    expect(result.errors).toBeUndefined();
    expect(result.data?.readBooks).not.toBe(null);
    expect(result.data?.readBooks.edges.length).toBe(6);
    expect(sanitizeTestObject(result.data)).toMatchSnapshot();
  });
});
