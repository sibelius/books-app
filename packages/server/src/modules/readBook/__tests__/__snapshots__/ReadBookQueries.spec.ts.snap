// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReadBook queries should get read book from node interface 1`] = `
Object {
  "readBook": Object {
    "book": Object {
      "name": "name 1",
    },
    "id": "FROZEN-ID",
    "readPages": 1,
  },
}
`;

exports[`ReadBook queries should query a connection of finished read books 1`] = `
Object {
  "readBooks": Object {
    "edges": Array [
      Object {
        "node": Object {
          "book": Object {
            "name": "name 13",
          },
          "id": "FROZEN-ID",
          "readPages": 10,
        },
      },
      Object {
        "node": Object {
          "book": Object {
            "name": "name 12",
          },
          "id": "FROZEN-ID",
          "readPages": 9,
        },
      },
      Object {
        "node": Object {
          "book": Object {
            "name": "name 11",
          },
          "id": "FROZEN-ID",
          "readPages": 8,
        },
      },
      Object {
        "node": Object {
          "book": Object {
            "name": "name 10",
          },
          "id": "FROZEN-ID",
          "readPages": 7,
        },
      },
      Object {
        "node": Object {
          "book": Object {
            "name": "name 9",
          },
          "id": "FROZEN-ID",
          "readPages": 6,
        },
      },
      Object {
        "node": Object {
          "book": Object {
            "name": "name 8",
          },
          "id": "FROZEN-ID",
          "readPages": 5,
        },
      },
      Object {
        "node": Object {
          "book": Object {
            "name": "name 7",
          },
          "id": "FROZEN-ID",
          "readPages": 4,
        },
      },
      Object {
        "node": Object {
          "book": Object {
            "name": "name 6",
          },
          "id": "FROZEN-ID",
          "readPages": 3,
        },
      },
      Object {
        "node": Object {
          "book": Object {
            "name": "name 5",
          },
          "id": "FROZEN-ID",
          "readPages": 2,
        },
      },
      Object {
        "node": Object {
          "book": Object {
            "name": "name 4",
          },
          "id": "FROZEN-ID",
          "readPages": 1,
        },
      },
    ],
  },
}
`;

exports[`ReadBook queries should query a connection of read books 1`] = `
Object {
  "readBooks": Object {
    "edges": Array [
      Object {
        "node": Object {
          "book": Object {
            "name": "name 1",
          },
          "id": "FROZEN-ID",
          "readPages": 6,
        },
      },
      Object {
        "node": Object {
          "book": Object {
            "name": "name 1",
          },
          "id": "FROZEN-ID",
          "readPages": 5,
        },
      },
      Object {
        "node": Object {
          "book": Object {
            "name": "name 1",
          },
          "id": "FROZEN-ID",
          "readPages": 4,
        },
      },
      Object {
        "node": Object {
          "book": Object {
            "name": "name 1",
          },
          "id": "FROZEN-ID",
          "readPages": 3,
        },
      },
      Object {
        "node": Object {
          "book": Object {
            "name": "name 1",
          },
          "id": "FROZEN-ID",
          "readPages": 2,
        },
      },
    ],
  },
}
`;

exports[`ReadBook queries should query a connection of read books with orderBy filter 1`] = `
Object {
  "readBooks": Object {
    "edges": Array [
      Object {
        "node": Object {
          "book": Object {
            "name": "name 1",
          },
          "id": "FROZEN-ID",
          "readPages": 5,
        },
      },
      Object {
        "node": Object {
          "book": Object {
            "name": "name 1",
          },
          "id": "FROZEN-ID",
          "readPages": 3,
        },
      },
      Object {
        "node": Object {
          "book": Object {
            "name": "name 1",
          },
          "id": "FROZEN-ID",
          "readPages": 200,
        },
      },
    ],
  },
}
`;

exports[`ReadBook queries should query a connection of unfinished read books 1`] = `
Object {
  "readBooks": Object {
    "edges": Array [
      Object {
        "node": Object {
          "book": Object {
            "name": "name 6",
          },
          "id": "FROZEN-ID",
          "readPages": 6,
        },
      },
      Object {
        "node": Object {
          "book": Object {
            "name": "name 5",
          },
          "id": "FROZEN-ID",
          "readPages": 5,
        },
      },
      Object {
        "node": Object {
          "book": Object {
            "name": "name 4",
          },
          "id": "FROZEN-ID",
          "readPages": 4,
        },
      },
      Object {
        "node": Object {
          "book": Object {
            "name": "name 3",
          },
          "id": "FROZEN-ID",
          "readPages": 3,
        },
      },
      Object {
        "node": Object {
          "book": Object {
            "name": "name 2",
          },
          "id": "FROZEN-ID",
          "readPages": 2,
        },
      },
      Object {
        "node": Object {
          "book": Object {
            "name": "name 1",
          },
          "id": "FROZEN-ID",
          "readPages": 1,
        },
      },
    ],
  },
}
`;
